### ISSUE TITLE
Adding types to existing Lua modules & global variables?

### ISSUE THREAD

**Issue by pdesaulniers** [hishamhm/tl#28]
Will tl support something equivalent to TypeScript's declaration files?

In LÖVE, all of the API is exposed through a global `love` table. I would like to declare all the functions in this table so that I can call them in a type-safe manner.

I would like to do the same for existing Lua libraries as well (such as Penlight).

**Comment by hishamhm** [comment]
Yes, I have thought about adding that at some point! This might just give me the push to prioritize this. :)

**Comment by hishamhm** [comment]
@pdesaulniers how about #30 for an initial implementation of this?

Please note that for now all types need to be `global` so they can be used across modules (a limitation noted in #29 which I should fix soon).

**Comment by pdesaulniers** [comment]
Awesome :) The definition files appear to work.

However, what if we want to declare the fields in a global created using Lua's C API? To achieve this, I think we'd have to use an empty .lua file:

application.d.tl:
```lua
global application = record
	quit: function()
end
```

application.lua:
```lua
-- Empty file; the `application` global is created using lua_setglobal in C code
```

main.tl:
```lua
require("application") -- Without the empty `application.lua` file, this call would fail
application.quit()
```

This is a bit weird, but I don't know if there's a better way of handling this.

**Comment by hishamhm** [comment]
@pdesaulniers this thought crossed my mind too. One solution would be something like:

```
tl -llove main.tl
```

This would pre-load `love.d.tl` when type checking, similar to `-l` in the `lua` interpreter. This would not be too hard to implement, it would be very suitable as a first contribution to the project — would you like to give this a try? 

(Of course, passing `-l` every time could get annoying, so perhaps we could have something like tsconfig for that purpose? (`tlconfig.lua`? :) )

**Comment by hishamhm** [comment]
I merged #30, so I think this particular issue can be closed! Let's open another one to track the question you raised [here](https://github.com/hishamhm/tl/issues/28#issuecomment-580051178)?

**Comment by ildar** [comment]
should be reflected in docs
